Invalid config if unix socket path too long

Tor fails to start if using Unix sockets where the path is too long. If consumer isn't logging warnings via a listener, then all that is seen is the exception that gets thrown when control port file times out.

This is problematic, as Unix Sockets are defaulted to for the control port if Unix sockets are available (android, Java 16+, or with the Unix domain socket extension lib) and library consumer did not declare any control port in their config.

Need to rework the Port Validation step to check path length of UnixSocket types (for hidden services as well).

In the mean time, consumers that are not setting a control port in their config and are using defaults can override that behavior and simply declare `Setting.Ports.Control` in their config.

`unixSocket.value?.value?.value` must be less than or equal to `105`, otherwise the following exception is thrown.

```
Caused by: java.net.SocketException: Unix domain path too long
```

--- a/library/manager/kmp-tor-manager/src/commonMain/kotlin/io/matthewnelson/kmp/tor/manager/internal/util/-PortValidator.kt
+++ b/library/manager/kmp-tor-manager/src/commonMain/kotlin/io/matthewnelson/kmp/tor/manager/internal/util/-PortValidator.kt
@@ -23,6 +23,7 @@ import io.matthewnelson.kmp.tor.controller.common.config.TorConfig.Setting.Ports
 import io.matthewnelson.kmp.tor.controller.common.config.TorConfig.Setting.UnixSockets
 import io.matthewnelson.kmp.tor.controller.common.file.Path
 import io.matthewnelson.kmp.tor.controller.common.internal.PlatformUtil
+import io.matthewnelson.kmp.tor.manager.common.exceptions.TorManagerException
 
 internal class PortValidator internal constructor(private val dataDir: Path) {
 
@@ -46,7 +47,12 @@ internal class PortValidator internal constructor(private val dataDir: Path) {
         }
     }
 
+    @Throws(TorManagerException::class)
     fun add(unixSocket: UnixSockets) {
+        if (!unixSocket.isAcceptablePathLength) {
+            throw TorManagerException("UnixSocket path must be less than or equal to 105 characters: $unixSocket")
+        }
+
         if (!unixSockets.add(unixSocket)) {
             return
         }
@@ -93,13 +99,20 @@ internal class PortValidator internal constructor(private val dataDir: Path) {
             // Prefer using unix domain socket if it's supported.
             @OptIn(InternalTorApi::class)
             val control = if (PlatformUtil.hasControlUnixDomainSocketSupport) {
-                UnixSockets.Control().set(
+                val unix = UnixSockets.Control()
+                unix.set(
                     TorConfig.Option.FileSystemFile(
                         dataDir.builder {
                             addSegment(UnixSockets.Control.DEFAULT_NAME)
                         }
                     )
                 )
+
+                if (unix.isAcceptablePathLength) {
+                    unix
+                } else {
+                    Ports.Control().set(AorDorPort.Auto)
+                }
             } else {
                 Ports.Control().set(AorDorPort.Auto)
             }
@@ -117,3 +130,6 @@ internal class PortValidator internal constructor(private val dataDir: Path) {
         return validated
     }
 }
+
+@Suppress("NOTHING_TO_INLINE", "KotlinRedundantDiagnosticSuppress")
+private inline val UnixSockets.isAcceptablePathLength: Boolean get() = (value?.value?.length ?: 0) <= 105

--- a/library/manager/kmp-tor-manager/src/commonTest/kotlin/io/matthewnelson/kmp/tor/manager/internal/util/PortValidatorUnitTest.kt
+++ b/library/manager/kmp-tor-manager/src/commonTest/kotlin/io/matthewnelson/kmp/tor/manager/internal/util/PortValidatorUnitTest.kt
@@ -117,6 +117,8 @@ class PortValidatorUnitTest {
 
         val socks = UnixSockets.Socks()
         socks.set(TorConfig.Option.FileSystemFile(Path(sb.toString())))
+        if (socks.value == null) return // skip test
+
         println(socks.value?.value?.length)
         validator.add(socks)
         sb.append('j')

--- a/library/manager/kmp-tor-manager/src/commonTest/kotlin/io/matthewnelson/kmp/tor/manager/internal/util/PortValidatorUnitTest.kt
+++ b/library/manager/kmp-tor-manager/src/commonTest/kotlin/io/matthewnelson/kmp/tor/manager/internal/util/PortValidatorUnitTest.kt
@@ -23,6 +23,7 @@ import io.matthewnelson.kmp.tor.controller.common.config.TorConfig.Setting.UnixS
 import io.matthewnelson.kmp.tor.controller.common.config.TorConfig.Option.AorDorPort
 import io.matthewnelson.kmp.tor.controller.common.file.Path
 import io.matthewnelson.kmp.tor.controller.common.internal.PlatformUtil
+import io.matthewnelson.kmp.tor.manager.common.exceptions.TorManagerException
 import kotlin.test.*
 
 @OptIn(InternalTorApi::class)
@@ -106,5 +107,23 @@ class PortValidatorUnitTest {
         val validated = validator.validate { true }
         assertTrue(validated.filterIsInstance<Ports.Socks>().isEmpty())
         assertFalse(validated.filterIsInstance<UnixSockets.Socks>().isEmpty())
+    }
+
+    @Test
+    fun givenUnixSocket_whenInvalidLength_thenFails() {
+        val sb = StringBuilder()
+        sb.append('/') // so check for unix fs succeeds
+        repeat(104) { sb.append('a') }
+
+        val socks = UnixSockets.Socks()
+        socks.set(TorConfig.Option.FileSystemFile(Path(sb.toString())))
+        println(socks.value?.value?.length)
+        validator.add(socks)
+        sb.append('j')
+        socks.set(TorConfig.Option.FileSystemFile(Path(sb.toString())))
+
+        assertFailsWith<TorManagerException> {
+            validator.add(socks)
+        }
     }
 }

--- a/library/manager/kmp-tor-manager/src/jvmAndroidMain/kotlin/io/matthewnelson/kmp/tor/manager/KmpTorLoader.kt
+++ b/library/manager/kmp-tor-manager/src/jvmAndroidMain/kotlin/io/matthewnelson/kmp/tor/manager/KmpTorLoader.kt
@@ -168,8 +168,13 @@ actual abstract class KmpTorLoader(protected val provider: TorConfigProvider) {
 
         torJob?.cancel()
 
-        val validated: TorConfigProvider.ValidatedTorConfig = withContext(dispatcher) {
-            provider.retrieve(excludeSettings) { port -> PortUtil.isTcpPortAvailable(port) }
+        val validated: TorConfigProvider.ValidatedTorConfig = try {
+            withContext(dispatcher) {
+                provider.retrieve(excludeSettings) { port -> PortUtil.isTcpPortAvailable(port) }
+            }
+        } catch (t: TorManagerException) {
+            torDispatcher.close()
+            return Result.failure(t)
         }
 
         val mkdirsFailure: TorManagerException? = withContext(dispatcher) {

